sampler.go

```diff
@@ -294,20 +301,45 @@ func newAdaptiveSampler(strategies *sampling.PerOperationSamplingStrategies, max
 		)
 		samplers[strategy.Operation] = sampler
 	}
-	return &adaptiveSampler{
+	return &AdaptiveSampler{
 		samplers:       samplers,
 		defaultSampler: newProbabilisticSampler(strategies.DefaultSamplingProbability),
 		lowerBound:     strategies.DefaultLowerBoundTracesPerSecond,
 		maxOperations:  maxOperations,
 	}
 }
 
-func (s *adaptiveSampler) IsSampled(id TraceID, operation string) (bool, []Tag) {
+func (s *AdaptiveSampler) IsSampled(id TraceID, operation string) (bool, []Tag) {
+	return false, nil
+}
+
+func (s *AdaptiveSampler) OnCreateSpan(span *Span) SamplingDecision {
+	operationName := span.OperationName()
+	samplerV1 := s.getSamplerForOperation(operationName)
+	sampled, tags := samplerV1.IsSampled(span.context.TraceID(), operationName)
+	return SamplingDecision{sample: sampled, retryable: true, tags: tags}
+}
+
+func (s *AdaptiveSampler) OnSetOperationName(span *Span, operationName string) SamplingDecision {
+	samplerV1 := s.getSamplerForOperation(operationName)
+	sampled, tags := samplerV1.IsSampled(span.context.TraceID(), operationName)
+	return SamplingDecision{sample: sampled, retryable: false, tags: tags}
```

Comment:
This is new behavior - do we require this right now?

Reply:
You mean new to Go client?  

I don't think we can avoid it. The main change is not actually here, but in OnCreateSpan, which *must* return retryable=true in order for tag sampling to work (otherwise the state will be finalized). So if we already have that in OnCreateSpan, adding OnSetOperationName() doesn't change things that much, imo.